contracts/KyberOasisReserve.sol

```diff
@@ -0,0 +1,273 @@
+    bool public tradeEnabled;
+    uint public feeBps;
+
+    function KyberOasisReserve(
+        address _kyberNetwork,
+        OtcInterface _otc,
+        TokenInterface _wethToken,
+        ERC20 _tradeToken,
+        address _admin,
+        uint _feeBps
+    ) public {
+        require(_admin != address(0));
+        require(_kyberNetwork != address(0));
+        require(_otc != address(0));
+        require(_wethToken != address(0));
+        require(_tradeToken != address(0));
+        require(_feeBps < 10000);
+        require(getDecimals(_wethToken) == MAX_DECIMALS);
+        require(getDecimals(_tradeToken) == MAX_DECIMALS);
+
+        kyberNetwork = _kyberNetwork;
+        otc = _otc;
+        wethToken = _wethToken;
+        tradeToken = _tradeToken;
+        admin = _admin;
+        feeBps = _feeBps;
+        tradeEnabled = true;
+
+        wethToken.approve(otc, 2**255);
+        tradeToken.approve(otc, 2**255);
+    }
+
+    event DepositToken(ERC20 token, uint amount);
+
+    /* solhint-disable no-complex-fallback */
+    function() public payable {
+        require(msg.sender == address(wethToken));
+        DepositToken(ETH_TOKEN_ADDRESS, msg.value);
+    }
+    /* solhint-enable no-complex-fallback */
+
+    event TradeExecute(
+        address indexed origin,
+        address src,
+        uint srcAmount,
+        address destToken,
+        uint destAmount,
+        address destAddress
+    );
+
+    function trade(
+        ERC20 srcToken,
+        uint srcAmount,
+        ERC20 destToken,
+        address destAddress,
+        uint conversionRate,
+        bool validate
+    )
+        public
+        payable
+        returns(bool)
+    {
+
+        require(tradeEnabled);
+        require(msg.sender == kyberNetwork);
+
+        require(doTrade(srcToken, srcAmount, destToken, destAddress, conversionRate, validate));
+
+        return true;
+    }
+
+    event TradeEnabled(bool enable);
+
+    function enableTrade() public onlyAdmin returns(bool) {
+        tradeEnabled = true;
+        TradeEnabled(true);
+
+        return true;
+    }
+
+    function disableTrade() public onlyAlerter returns(bool) {
+        tradeEnabled = false;
+        TradeEnabled(false);
+
+        return true;
+    }
+
+    event KyberNetworkSet(address kyberNetwork);
+
+    function setKyberNetwork(address _kyberNetwork) public onlyAdmin {
+        require(_kyberNetwork != address(0));
+
+        kyberNetwork = _kyberNetwork;
+        KyberNetworkSet(kyberNetwork);
+    }
+
+    event OtcSet(address otc);
+
+    function setOtc(OtcInterface _otc) public onlyAdmin {
+        require(_otc != address(0));
+
+        wethToken.approve(otc, 0);
+        tradeToken.approve(otc, 0);
+        wethToken.approve(_otc, 2**255);
+        tradeToken.approve(_otc, 2**255);
+
+        otc = _otc;
+        OtcSet(otc);
+    }
+
+    event WethTokenSet(TokenInterface wethToken);
+
+    function setWethToken(TokenInterface _wethToken) public onlyAdmin {
```

Comment:
maybe remove it? these things are pretty stable. worse case we redeploy the reserve. it does not hold funds anw.

Reply:
done.